// @inrupt/solid-client-authn-node/dist/login/oidc/__mocks__/IOidcOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockOidcOptions = exports.mockDefaultOidcOptions = exports.standardOidcOptions = void 0;
exports.standardOidcOptions = {
    sessionId: "mySession",
    issuer: "https://example.com",
    dpop: true,
    redirectUrl: "https://app.example.com",
    handleRedirect: jest.fn((url) => url),
    issuerConfiguration: {
        issuer: "https://example.com",
        authorizationEndpoint: "https://example.com/auth",
        tokenEndpoint: "https://example.com/token",
        jwksUri: "https://example.com/jwks",
        subjectTypesSupported: [],
        claimsSupported: [],
    },
    client: {
        clientId: "coolApp",
    },
};
const mockDefaultOidcOptions = () => {
    return { ...exports.standardOidcOptions };
};
exports.mockDefaultOidcOptions = mockDefaultOidcOptions;
const mockOidcOptions = (overriddenOptions) => {
    return {
        ...exports.standardOidcOptions,
        ...overriddenOptions,
    };
};
exports.mockOidcOptions = mockOidcOptions;
