// @comunica/actor-query-operation-update-drop/lib/ActorQueryOperationDrop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationDrop = void 0;
const bus_query_operation_1 = require("@comunica/bus-query-operation");
const rdf_data_factory_1 = require("rdf-data-factory");
const DF = new rdf_data_factory_1.DataFactory();
/**
 * A [Query Operation](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation) actor
 * that handles SPARQL drop operations.
 */
class ActorQueryOperationDrop extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'drop');
    }
    async testOperation(pattern, context) {
        bus_query_operation_1.ActorQueryOperation.throwOnReadOnly(context);
        return true;
    }
    async runOperation(pattern, context) {
        // Delegate to update-quads bus
        const { updateResult } = await this.mediatorUpdateQuads.mediate({
            deleteGraphs: {
                graphs: pattern.source === 'DEFAULT' ? DF.defaultGraph() : pattern.source,
                requireExistence: !pattern.silent,
                dropGraphs: true,
            },
            context,
        });
        return {
            type: 'update',
            updateResult,
        };
    }
}
exports.ActorQueryOperationDrop = ActorQueryOperationDrop;
