// @solid-rest/file/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SolidRestFile = void 0;

var _path = _interopRequireDefault(require("path"));

var _stream = require("stream");

var _core = _interopRequireDefault(require("@solid-rest/core"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _mimeTypes = _interopRequireDefault(require("mime-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import concatStream from "concat-stream";
class SolidRestFile {
  /**
   * file system backend for Solid-Rest
   * @constructor
   * @return {prefix:"file",name:"solid-rest-file-version"}
   */
  constructor() {
    this.prefix = "file";
    this.name = "solid-rest-file-2.0.0";
    return new _core.default({
      plugin: this
    });
  }
  /**
   * check if file exists
   * @param filePath
   * @return {boolean}
   */


  async itemExists(path) {
    return _fsExtra.default.existsSync(path);
  }
  /**
   * check if thing is Container or Resource
   * @param filePath
   * @return "Container" | "Resource" | null
   */


  async itemType(path) {
    let stat;

    try {
      stat = await _fsExtra.default.lstatSync(path);
    } catch (err) {}

    if (!stat) return null;
    if (stat.isSymbolicLink()) return null;
    if (stat.isDirectory()) return "Container";
    if (stat.isFile()) return "Resource";
    return null;
  }
  /**
   * read a folder
   * @param filePath
   * @return on success : a possibly empty array of child fileNames (not paths)
   * @return on failure : false
   */


  async getContainer(pathname) {
    let files;
    let newFiles = [];

    try {
      files = await _fsExtra.default.readdirSync(pathname);

      for (var f in files) {
        newFiles.push(await this.getItemInfo(pathname + files[f]));
      }
    } catch (e) {
      console.log(e);
      return false;
    }

    return newFiles;
  }
  /**
   * examine a requested item
   * @param filePath, request object
   * @return item object
   */


  async getItemInfo(fn) {
    fn = fn.replace(/^file:\/\//, '');

    let mimetype = _mimeTypes.default.lookup(fn); // mimetype from ext


    let type = await this.itemType(fn); // Container/Resource

    let exists = await this.itemExists(fn);
    if (!type && fn.endsWith('/')) type = "Container";
    let read = true,
        write = true;

    if (exists) {
      try {
        _fsExtra.default.accessSync(fn, _fsExtra.default.constants.R_OK);
      } catch (e) {
        read = false;
      }

      try {
        _fsExtra.default.accessSync(fn, _fsExtra.default.constants.W_OK);
      } catch (e) {
        write = false;
      }
    }

    let mode = {
      read: read,
      write: write,
      append: write,
      control: write
    };
    let item = {
      fileName: fn,
      path: fn,
      extension: _mimeTypes.default.extension(fn),
      mode: mode,
      exists: exists,
      isContainer: type === "Container" ? true : false,
      mimetype: mimetype
    };
    return Promise.resolve(item);
  }
  /**
   * read a file
   * @param filePath
   * @return on success : body | Response object
   * @return on failure false  | Response object
   */


  async getResource(pathname) {
    let ctype, encoding, bodyData;

    try {
      ctype = _mimeTypes.default.contentType(pathname);
      encoding = ctype.match(/text|application/) ? "string" : null;
      bodyData = encoding ? await _fsExtra.default.readFile(pathname, encoding) : await _fsExtra.default.readFile(pathname);
    } catch (e) {
      console.log("Error" + e);
      return false;
    }

    return bodyData;
  }
  /**
   * write a file
   * @param filePath, content
   * @return true | Response object on success
   * @return false | Response object on failure
   */


  async putResource(pathname, content, ctype) {
    let successCode = true;
    let failureCode = false;
    return new Promise(async resolve => {
      let writeIt = false;
      if (typeof content === "undefined") content = "";

      if (typeof content === "string") {
        writeIt = true;
      } else if (content.stream) {
        content = await content.stream();
        content = await content.read();
        writeIt = true;
      } else if (content.text) {
        content = await content.text();
        writeIt = true;
      }

      if (writeIt) {
        try {
          await _fsExtra.default.writeFileSync(pathname, content);
          return resolve(successCode);
        } catch (e) {
          console.log(e);
          return resolve(failureCode);
        }
      }

      if (!content.pipe && typeof FileReader != "undefined") {
        var fileReader = new FileReader();

        fileReader.onload = function () {
          _fsExtra.default.writeFileSync(pathname, Buffer.from(new Uint8Array(this.result)));
        };

        fileReader.onloadend = () => {
          return resolve(successCode);
        };

        fileReader.onerror = err => {
          console.log(err);
          return resolve(failureCode);
        };

        fileReader.readAsArrayBuffer(content);
      } else {
        content = content || "";
        content = this._makeStream(content);
        content.pipe(_fsExtra.default.createWriteStream(pathname)).on('finish', () => {
          return resolve(successCode);
        }).on('error', err => {
          console.log(err);
          return resolve(failureCode);
        });
      }
    });
  }
  /**
   * delete file
   * @param filePath
   * @return true on success, false on failure
   */


  async deleteResource(fn) {
    return new Promise(function (resolve) {
      _fsExtra.default.unlink(fn, function (err) {
        if (err) resolve(false);else resolve(true);
      });
    });
  }
  /**
   * delete folder
   * @param filePath
   * @return true on success, false on failure
   */


  deleteContainer(fn) {
    return new Promise(async function (resolve) {
      await _fsExtra.default.rmdir(fn, function (err) {
        if (err) {
          resolve(false);
        } else {
          resolve(true);
        }
      });
    });
  }
  /**
   * create folder
   * @param filePath
   * @return true on success, false on failure
   */


  async postContainer(fn) {
    fn = fn.replace(/\/$/, '');
    return new Promise(function (resolve) {
      if (_fsExtra.default.existsSync(fn)) {
        return resolve(true);
      }

      _fsExtra.default.mkdirp(fn, {}, err => {
        if (err) {
          return resolve(false);
        } else {
          return resolve(true);
        }
      });
    });
  }
  /**
   * create parent and intermediate folders
   * @param filePath
   * @return true on success, false on failure
   */


  async makeContainers(pathname) {
    const foldername = _path.default.dirname(pathname);

    if (!_fsExtra.default.existsSync(foldername)) {
      try {
        _fsExtra.default.mkdirpSync(foldername);

        return Promise.resolve(true);
      } catch {
        return Promise.resolve(false);
      }
    }

    return Promise.resolve(true);
  }
  /**
   * creates a ReadableStream
   * @param content as text or streamable object
   * @return on success : a ReadableStream
   * @return on failure : false
   */


  _makeStream(text) {
    if (typeof text === 'object' && typeof text.stream === 'function') {
      return text.stream();
    }

    let s = new _stream.Readable();
    s.push(text);
    s.push(null);
    return s;
  }

} // end of SolidRestFile

/**
 * return parent url with / at the end.
 * If no parent exists return null
 * @param {string} url 
 * @returns {string|null}
*/


exports.SolidRestFile = SolidRestFile;

function getParent(url) {
  while (url.endsWith('/')) url = url.slice(0, -1);

  if (!url.includes('/')) return null;
  return url.substring(0, url.lastIndexOf('/')) + '/';
} // ENDS
